From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Prathamesh Shete <pshete@nvidia.com>
Date: Fri, 18 Oct 2024 06:14:21 +0000
Subject: [PATCH 029/184] NVIDIA: SAUCE: gpio: tegra186: Add suspend resume
 support

Squash following 2 commits needed for suspend resume support in GPIO
driver:
 - gpio-tegra: add suspend resume support for gpio
   (9af7edc30a49e309875d8b022d4b5ad8d727b027)
 - GPIO: Add check for VM access before reading the register
   (3bbe623a714e54ebd632964010d322f086f8e241)

Bug 5143785
Bug 5250926

Signed-off-by: Prathamesh Shete <pshete@nvidia.com>
Change-Id: I21f049960b05198801d2e570f1b6061d0e037a15
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3474939
Reviewed-by: Laxman Dewangan <ldewangan@nvidia.com>
(cherry picked from commit 342d475fbd30f1e16aadce938b462b2854a84b61)

diff --git a/drivers/gpio/gpio-tegra186.c b/drivers/gpio/gpio-tegra186.c
index fb26402b6c47..eeaee6cb44b8 100644
--- a/drivers/gpio/gpio-tegra186.c
+++ b/drivers/gpio/gpio-tegra186.c
@@ -122,6 +122,13 @@ struct tegra_gpio_soc {
 	bool has_vm_support;
 };
 
+struct tegra_gpio_regs {
+	u32 val;
+	u32 conf;
+	u32 out;
+	u32 debounce;
+};
+
 struct tegra_gpio {
 	struct gpio_chip gpio;
 	unsigned int num_irq;
@@ -133,6 +140,7 @@ struct tegra_gpio {
 
 	void __iomem *secure;
 	void __iomem *base;
+	struct tegra_gpio_regs *gpio_rval;
 };
 
 static const struct tegra_gpio_port *
@@ -889,6 +897,10 @@ static int tegra186_gpio_probe(struct platform_device *pdev)
 	if (IS_ERR(gpio->base))
 		return PTR_ERR(gpio->base);
 
+	gpio->gpio_rval = devm_kzalloc(&pdev->dev, gpio->soc->num_ports * 8 *
+			sizeof(*gpio->gpio_rval), GFP_KERNEL);
+	if (!gpio->gpio_rval)
+		return -ENOMEM;
 	err = platform_irq_count(pdev);
 	if (err < 0)
 		return err;
@@ -1037,9 +1049,69 @@ static int tegra186_gpio_probe(struct platform_device *pdev)
 		offset += port->pins;
 	}
 
+	platform_set_drvdata(pdev, gpio);
 	return devm_gpiochip_add_data(&pdev->dev, &gpio->gpio, gpio);
 }
 
+#ifdef CONFIG_PM_SLEEP
+static int tegra_gpio_suspend_late(struct device *dev)
+{
+	struct tegra_gpio *gpio = dev_get_drvdata(dev);
+	struct tegra_gpio_regs *regs;
+	void __iomem *base;
+	int i;
+
+	for (i = 0; i < gpio->gpio.ngpio; i++) {
+		if (!tegra186_gpio_is_accessible(gpio, i))
+			continue;
+
+		base = tegra186_gpio_get_base(gpio, i);
+		if (WARN_ON(base == NULL))
+			return -EINVAL;
+		regs = &gpio->gpio_rval[i];
+		regs->conf = readl(base + TEGRA186_GPIO_ENABLE_CONFIG),
+		regs->out = readl(base + TEGRA186_GPIO_OUTPUT_CONTROL),
+		regs->val = readl(base + TEGRA186_GPIO_OUTPUT_VALUE),
+		regs->debounce = readl(base + TEGRA186_GPIO_DEBOUNCE_CONTROL);
+	}
+		return 0;
+
+}
+
+static int tegra_gpio_resume_noirq(struct device *dev)
+{
+	struct tegra_gpio *gpio = dev_get_drvdata(dev);
+	struct tegra_gpio_regs *regs;
+	void __iomem *base;
+	int i;
+
+	for (i = 0; i < gpio->gpio.ngpio; i++) {
+		if (!tegra186_gpio_is_accessible(gpio, i))
+				continue;
+
+		base = tegra186_gpio_get_base(gpio, i);
+		if (WARN_ON(base == NULL))
+			return -EINVAL;
+		regs = &gpio->gpio_rval[i];
+
+		writel(regs->val,  base + TEGRA186_GPIO_OUTPUT_VALUE);
+		writel(regs->out,  base + TEGRA186_GPIO_OUTPUT_CONTROL);
+		writel(regs->conf, base + TEGRA186_GPIO_ENABLE_CONFIG);
+		writel(regs->debounce, base + TEGRA186_GPIO_DEBOUNCE_CONTROL);
+	}
+
+	return 0;
+}
+
+static const struct dev_pm_ops tegra_gpio_pm = {
+	.suspend_late = tegra_gpio_suspend_late,
+	.resume_noirq = tegra_gpio_resume_noirq,
+};
+#define TEGRA_GPIO_PM		(&tegra_gpio_pm)
+#else
+#define TEGRA_GPIO_PM		NULL
+#endif /* CONFIG_PM_SLEEP */
+
 #define TEGRA_GPIO_PORT(_prefix, _name, _bank, _port, _pins) \
 	[_prefix##_GPIO_PORT_##_name] = { \
 		.name = #_name, \
@@ -1487,6 +1559,7 @@ static struct platform_driver tegra186_gpio_driver = {
 		.name = "tegra186-gpio",
 		.of_match_table = tegra186_gpio_of_match,
 		.acpi_match_table = tegra186_gpio_acpi_match,
+		.pm = TEGRA_GPIO_PM,
 	},
 	.probe = tegra186_gpio_probe,
 };
-- 
2.55.0

